research/picai/prepare_data.py

```diff
@@ -0,0 +1,209 @@
+        annotation_array = sitk.GetArrayFromImage(annotation)
+        case_label = float(np.max(annotation_array))
+        case_labels.append(case_label)
+    return case_labels
+
+
+def generate_dataset_json(
+    paths_for_each_sample: Sequence[Tuple[Sequence[Path], Path]],
+    write_dir: Path,
+    splits_path: Optional[Path] = None,
+) -> None:
+
+    labels = get_labels(paths_for_each_sample)
+
+    # Since we are storing paths inside a json file, we need to convert to strings to be serializable.
+    # Also pack in label for each samples in the tuple containing scan paths and annotation path.
+    scan_annotation_label_list = [
+        (list(map(str, tup[0])), str(tup[1]), lbl) for tup, lbl in zip(paths_for_each_sample, labels)
+    ]
+
+    if splits_path is None:
+        # If splits_path is None, create a singe dataset overview
+        d = {}
+        d["image_paths"], d["label_paths"], d["case_label"] = zip(*scan_annotation_label_list)
+        write_path = os.path.join(write_dir, "train-fold-all.json")
+        with open(write_path, "w") as f:
+            json.dump(d, f)
+    else:
+        # If splits_path is not None, create a dataset overview for each split
+        with open(splits_path, "r") as splits_f:
+            splits = json.load(splits_f)
+
+        for i, split in enumerate(splits):
+            # Add image (scan) and label (annotation) paths that include any subject ids from split
+            # For both train and val
+            train_d, val_d = {}, {}
+            train_list = [
+                tup
+                for tup in scan_annotation_label_list
+                if any([subject_id in str(tup[1]) for subject_id in split["train"]])
+            ]
+            val_list = [
+                tup
+                for tup in scan_annotation_label_list
+                if any([subject_id in str(tup[1]) for subject_id in split["val"]])
+            ]
+            train_d["image_paths"], train_d["label_paths"], train_d["case_label"] = zip(*train_list)
+            val_d["image_paths"], val_d["label_paths"], val_d["case_label"] = zip(*val_list)
+
+            # Create path of json files and write to disk
+            train_write_path = os.path.join(write_dir, f"train-fold-{i}.json")
+            val_write_path = os.path.join(write_dir, f"val-fold-{i}.json")
+
+            with open(train_write_path, "w") as f:
+                json.dump(train_d, f)
+
+            with open(val_write_path, "w") as f:
+                json.dump(val_d, f)
+
+
+def preprare_data(
+    scans_read_dir: Path,
+    annotation_read_dir: Path,
+    scans_write_dir: Path,
+    annotation_write_dir: Path,
+    overview_write_dir: Path,
+    size: Optional[Sequence[int]] = None,
+    physical_size: Optional[Sequence[int]] = None,
+    spacing: Optional[Sequence[int]] = None,
+    scan_extension: str = "mha",
+    annotation_extension: str = ".nii.gz",
+    num_threads: int = 4,
+    splits_path: Optional[Path] = None,
+) -> None:
```

Comment:
Some documentation on these input args might be helpful (I know there are a lot and so it might be a pain though)

Reply:
I agree some documentation would be good hear :)